rust/rules.bzl

```diff
@@ -0,0 +1,117 @@
+load("@io_bazel_rules_rust//rust:rust.bzl", "rust_library")
+load("//protobuf:rules.bzl", "proto_compile", "proto_language_deps", "proto_repositories")
+load("//rust/raze:crates.bzl", "raze_fetch_remote_crates")
+
+def rust_proto_repositories(
+    lang_deps = {},
+    lang_requires = [],
+    **kwargs):
+  raze_fetch_remote_crates()
+
+def rust_proto_compile(langs = [str(Label("//rust"))], **kwargs):
+  proto_compile(langs = langs, **kwargs)
+
+PB_COMPILE_DEPS = [
+    "@org_pubref_rules_protobuf//rust/raze:protobuf",
+]
+
+GRPC_COMPILE_DEPS = PB_COMPILE_DEPS + [
+    "@org_pubref_rules_protobuf//rust/raze:grpc",
+    "@org_pubref_rules_protobuf//rust/raze:tls_api",
+    "@org_pubref_rules_protobuf//rust/raze:tls_api_stub",
+]
+
+def _basename(f):
+  return f.basename[:-len(f.extension) - 1]
+
+def _gen_lib_impl(ctx):
+  content = ["extern crate protobuf;"]
+  if ctx.attr.grpc:
+    content.append("extern crate grpc;")
+    content.append("extern crate tls_api;")
+  for dep in ctx.attr.deps:
+    content.append("extern crate %s;" % dep.label.name)
+    content.append("use %s::*;" % dep.label.name)
```

Comment:
Have you tested this with the protobuf file generator you're using? I was historically using stepancheg's rust proto library, and it uses crate-wide "allow" directives. These directives must appear before any use statement. To solve that, I just put the use statements after the content.

This may not be an issue for your particular proto library.

Reply:
I haven't seen any issue in the examples I am testing on, but I moved the use statement at the end.